File: season2_chapter3/python/time_series_query.py

from typing import Union
import pandas as pd
import polars as pl


def by_pandas(df: pd.DataFrame, window_size: int):
    """pandasを用いて移動平均を計算する"""
    features = df.rolling(window_size).mean()
    return features


def by_polars(df: Union[pl.DataFrame, pl.LazyFrame], window_size: int):
    """polarsを用いて移動平均を計算する"""
    features = df.select(
        [
            pl.col('value').rolling_mean(window_size),
        ]
    )
    return features


if __name__ == '__main__':
    # pandasでの特徴量計算
    series_by_pandas = by_pandas(pd.read_csv('../data/time_series.csv'), 50)
    print(series_by_pandas)
    # polarsでの特徴量計算(遅延評価しない)
    series_by_eager_polars = by_polars(pl.read_csv('../data/time_series.csv'), 50)
    print(series_by_eager_polars)
    # polarsでの特徴量計算(遅延評価する)
    series_by_lazy_polars = by_polars(pl.scan_csv('../data/time_series.csv'), 50).collect()
    print(series_by_lazy_polars)

File: season2_chapter3/rust/src/time_series_query.rs
use polars::prelude::{col, DataFrame, Duration, LazyCsvReader, LazyFrame, RollingOptions};

// プロジェクトディレクトリからの相対パスを絶対パスにするユーティリティ
fn get_csv_path(relative_path: &str) -> std::path::PathBuf {
    let project_path = env!("CARGO_MANIFEST_DIR");
    std::path::Path::new(project_path)
        .parent()
        .unwrap()
        .join(relative_path)
}

// polarsで移動平均を計算する
fn by_polars(df: LazyFrame, window_size: Duration) -> anyhow::Result<DataFrame> {
    let rolling_options = RollingOptions {
        window_size,
        ..RollingOptions::default()
    };
    let features = df
        .select([col("value").rolling_mean(rolling_options)])
        .collect()?;
    Ok(features)
}

fn main() -> anyhow::Result<()> {
    let csv_path = get_csv_path("data/time_series.csv");
    // csvを遅延読み込みする
    let df = LazyCsvReader::new(csv_path).has_header(true).finish()?;
    // polarsでの特徴量計算(遅延評価する)
    let features = by_polars(df, Duration::new(50))?;
    println!("{:?}", features);
    Ok(())
}
